File: Python/Linked-List/Remove_cycle_in_Linked_List.py
# Python program to detect and remove loop in linked list

class Node:

	def __init__(self, data):
		self.data = data
		self.next = None

class LinkedList:

	def __init__(self):
		self.head = None

	def detectAndRemoveLoop(self):
		slow_p = fast_p = self.head
		
		while(slow_p and fast_p and fast_p.next):
			slow_p = slow_p.next
			fast_p = fast_p.next.next

			if slow_p == fast_p:
				self.removeLoop(slow_p)
		
				# Return 1 to indicate that loop is found
				return 1
		
		# Return 0 to indicate that there is no loop
		return 0

	def removeLoop(self, loop_node):
		ptr1 = loop_node
		ptr2 = loop_node
		
		k = 1
		while(ptr1.next != ptr2):
			ptr1 = ptr1.next
			k += 1

		
		ptr1 = self.head
		
		ptr2 = self.head
		for i in range(k):
			ptr2 = ptr2.next

		while(ptr2 != ptr1):
			ptr1 = ptr1.next
			ptr2 = ptr2.next

		while(ptr2.next != ptr1):
			ptr2 = ptr2.next

		ptr2.next = None

	def push(self, new_data):
		new_node = Node(new_data)
		new_node.next = self.head
		self.head = new_node

	def printList(self):
		temp = self.head
		while(temp):
			print(temp.data, end = ' ')
			temp = temp.next


# Driver program
llist = LinkedList()
llist.push(10)
llist.push(4)
llist.push(15)
llist.push(20)
llist.push(50)

# Create a loop for testing
llist.head.next.next.next.next.next = llist.head.next.next

llist.detectAndRemoveLoop()

print("Linked List after removing loop")
llist.printList()

File: C/Linked-List/Remove_cycle_in_Linked_List.c
// C program to remove a loop from the linked list  
#include <stdio.h>  
#include <stdlib.h>  
  
struct node  
{  
    int data;  
    struct node *next;  
};  
struct node *head;    
struct node* create_node(int value)  
{  
   // creating a temp variable of type node..   
   struct node *temp = (struct node*)malloc(sizeof(struct node));  
        temp->data = value;  
        temp->next = NULL;  
        return temp;  
   
          
}   
int detect_loop(struct node *ptr)  
{  
   struct node *fast;  
   struct node *slow; 
   fast = slow = ptr;  
   while (slow && fast && fast->next)  
   {  
       slow = slow->next;  
       fast = fast->next->next;  
      
    if(slow==fast)  
       {  
           removeloop(slow, head);   
           return 1;  
       }  
    }  
}  
  
void removeloop(struct node *slow, struct node *head)  
{  
    struct node *p1;  
    struct node *p2;  
    p1 = slow;  
    p2 = head;  
      
while(p1->next!=p2->next)  
    {  
        p1 = p1->next;  
        p2 = p2->next;  
    }  
    p1->next = NULL;  
      
}  
void display(struct node *temp)  
{  
   while (temp !=NULL)  
    {  
        printf("%d, ", temp->data);  
        temp = temp->next;  
    }  
}  
int main()  
{  
    head = create_node(50);  
    head->next = create_node(20);  
    head->next->next = create_node(15);  
    head->next->next->next = create_node(4);  
    head->next->next->next->next = create_node(10);  
   /* Create a loop for testing */  
    head->next->next->next->next->next = head->next->next;  
    detect_loop(head);  
    display(head);  
   return 0;  
} 